File: system_monitor/server/state.py
from __future__ import annotations

import asyncio
import threading
from typing import Any


class LiveHub:
    def __init__(self) -> None:
        self.connections: list[Any] = []
        self._loop: asyncio.AbstractEventLoop | None = None

    def bind_loop(self, loop: asyncio.AbstractEventLoop) -> None:
        self._loop = loop

    async def connect(self, websocket: Any) -> None:
        await websocket.accept()
        self.connections.append(websocket)

    def disconnect(self, websocket: Any) -> None:
        if websocket in self.connections:
            self.connections.remove(websocket)

    async def broadcast(self, payload: dict[str, Any]) -> None:
        dead: list[Any] = []
        for ws in self.connections:
            try:
                await ws.send_json(payload)
            except Exception:
                dead.append(ws)
        for ws in dead:
            self.disconnect(ws)

    def schedule_broadcast(self, payload: dict[str, Any]) -> None:
        if self._loop is None or not self.connections:
            return
        asyncio.run_coroutine_threadsafe(self.broadcast(payload), self._loop)


class FleetState:
    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._latest_by_agent: dict[str, dict[str, dict[str, Any]]] = {}

    def update_agent(self, agent_id: str, latest: dict[str, dict[str, Any]]) -> None:
        with self._lock:
            current = self._latest_by_agent.setdefault(agent_id, {})
            current.update(latest)

    def get_agent_snapshot(self, agent_id: str) -> dict[str, dict[str, Any]]:
        with self._lock:
            return dict(self._latest_by_agent.get(agent_id, {}))

    def get_combined_snapshot(self) -> dict[str, dict[str, Any]]:
        with self._lock:
            combined: dict[str, dict[str, Any]] = {}
            for agent_latest in self._latest_by_agent.values():
                combined.update(agent_latest)
            return combined

File: internal/hubserver/state.go
package hubserver

import (
	"encoding/json"
	"sync"

	"github.com/gorilla/websocket"
)

type LiveHub struct {
	mu          sync.Mutex
	connections map[*websocket.Conn]bool
}

func NewLiveHub() *LiveHub {
	return &LiveHub{connections: map[*websocket.Conn]bool{}}
}

func (h *LiveHub) Add(conn *websocket.Conn) {
	h.mu.Lock()
	h.connections[conn] = true
	h.mu.Unlock()
}

func (h *LiveHub) Remove(conn *websocket.Conn) {
	h.mu.Lock()
	delete(h.connections, conn)
	h.mu.Unlock()
}

func (h *LiveHub) Broadcast(payload map[string]interface{}) {
	data, err := json.Marshal(payload)
	if err != nil {
		return
	}
	h.mu.Lock()
	dead := make([]*websocket.Conn, 0)
	for conn := range h.connections {
		if err := conn.WriteMessage(websocket.TextMessage, data); err != nil {
			dead = append(dead, conn)
		}
	}
	for _, conn := range dead {
		delete(h.connections, conn)
		_ = conn.Close()
	}
	h.mu.Unlock()
}

func (h *LiveHub) ScheduleBroadcast(payload map[string]interface{}) {
	h.Broadcast(payload)
}

type FleetState struct {
	mu             sync.Mutex
	latestByAgent  map[string]map[string]map[string]interface{}
}

func NewFleetState() *FleetState {
	return &FleetState{latestByAgent: map[string]map[string]map[string]interface{}{}}
}

func (f *FleetState) UpdateAgent(agentID string, latest map[string]map[string]interface{}) {
	f.mu.Lock()
	current := f.latestByAgent[agentID]
	if current == nil {
		current = map[string]map[string]interface{}{}
		f.latestByAgent[agentID] = current
	}
	for k, v := range latest {
		current[k] = v
	}
	f.mu.Unlock()
}

func (f *FleetState) GetAgentSnapshot(agentID string) map[string]map[string]interface{} {
	f.mu.Lock()
	defer f.mu.Unlock()
	current := f.latestByAgent[agentID]
	out := make(map[string]map[string]interface{}, len(current))
	for k, v := range current {
		out[k] = v
	}
	return out
}

func (f *FleetState) RemoveAgent(agentID string) {
	f.mu.Lock()
	delete(f.latestByAgent, agentID)
	f.mu.Unlock()
}

func (f *FleetState) GetCombinedSnapshot() map[string]map[string]interface{} {
	f.mu.Lock()
	defer f.mu.Unlock()
	combined := map[string]map[string]interface{}{}
	for _, agentLatest := range f.latestByAgent {
		for k, v := range agentLatest {
			combined[k] = v
		}
	}
	return combined
}
